Handle guest logs

skycastlelily opened the pull request:
Fetch log content and save it to the workdir/provision/ List the log dir in guests.yaml

Pull Request Checklist

* [x] implement the feature
* [ ] write the documentation
* [ ] extend the test coverage
* [ ] update the specification
* [ ] adjust plugin docstring
* [ ] modify the json schema
* [ ] mention the version
* [ ] include a release note

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,34 @@ def pull(
+        """fetch and return content of a requested log"""
+        if log_name == 'dmesg':
+            log_content = self.execute(Command('dmesg')).stdout
+        if log_content:
```
If `log_name` is not `dmesg`, then at this moment, Python will blow up because `log_content` would be an unknown name. That's a bug, but we can avoid it by refactoring the code: do we even have to store the content? I'd say it's enough to return it right away:

```python
        if log_name == 'dmesg':
            return self.execute(Command('dmesg')).stdout
        return None
```

skycastlelily replied:
Thanks for your review^^,gonna to modify the code accordingly.

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,34 @@ def pull(
+            return log_content
+        return None
+
+    def store_log(self, log_name: str, log_content: str) -> Optional[str]:
```
The file path to save into must be the parameter: it's not up to the guest instance to decide where the log is supposed to end. It may be a file in step's workdir, it may be a file on NFS mount, or it may be `/dev/null`. The point is, the guest does not know the context. Accept file path and content, and save. Plus, there is not much to return from this method, `-> None` would be enough. Your code returns the path, but that should be an input rather than outcome of the call to `store_log()`.

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,34 @@ def pull(
+    def handle_guest_log(self, log_list: Optional[list[str]] = None) -> None:
+        """get log content and save it to the workdir/provision/,
+        list the log dir in guests.yaml"""
+        log_list = log_list if log_list else []
```
Short Python idiom: `log_list = log_list or []`.

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,34 @@ def pull(
+            return f'{self.workdir}/{log_name}'
+        return None
+
+    def handle_guest_log(self, log_list: Optional[list[str]] = None) -> None:
```
Encoding the type in the variable name is usually better to avoid, pure English is often better: `logs`. It's a plural, i.e. "it's a collection" aspect is clearly visible, and it's easy to catch while reading the code.

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,34 @@ def pull(
+            if log_content:
+                log_path = self.store_log(log_name, log_content)
+                if log_path:
+                    self.write(Path('guests.yaml'), f'{ log_name} : {log_path}', 'a')
```
We will need a better way how to expose logs. For now, I would leave this place empty, but we will extend the `Guest` class enough to take care of this. Note that `guests.yaml` may contain multiple guests, this simple approach might break the structure or append all logs to the last guest listed in the file.

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,43 @@ def pull(
+            if log_content:
+                self.store_log(log_name, self.workdir, log_content)
+            else:
+                self.debug(f'no content in {log_name}')
```
Why not make this the default implementation of `store_log()` and `handle_guest_logs()` in the `Guest` class? There is *nothing* that would be `local`-specific in these two methods, and they can provide a handy default behavior for all modules.

skycastlelily replied:
there is nothing local-specific in store_log,but what logs can be provided depends on different guests,I was thinking define a LOG_LIST for each plugin,so I put handle_guest_logs to local.py,but when I saw your comment,I suddenly realized that I could define a log_list property for each plugin,and put both of that two methods to Guest class,right?^^

happz replied:
Yep, "the list of logs" and "saving of the list of logs" are two distinct parts, and the plugin may easily override the former without changing a bit in the latter.

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,43 @@ def pull(
+    def store_log(
+            self,
+            log_name: Optional[str],
+            log_path: Optional[Path],
```
1. As we discussed, `log_path` is probably not something optional, it's a given input parameter. I can imagine that it might be a directory, then the method would save the log into this directory using `log_name`, but still, I see no use for `log_path` being unset. That would mean the guest implementation would need to decide the destination, and that's something we should avoid as it lacks context.
2. `log_name`, being optional, deserves a default value, i.e. `log_name: Optional[str] = None`.

skycastlelily replied:
> 1. As we discussed, `log_path` is probably not something optional, it's a given input parameter. I can imagine that it might be a directory, then the method would save the log into this directory using `log_name`, but still, I see no use for `log_path` being unset. That would mean the guest implementation would need to decide the destination, and that's something we should avoid as it lacks context.

Got it, I was thinking maybe users want to store the logs to dir they ran tmt command:)

> 
>     2. `log_name`, being optional, deserves a default value, i.e. `log_name: Optional[str] = None`.

Yeah,will updated accordingly

happz replied:
> Got it, I was thinking maybe users want to store the logs to dir they ran tmt command:)

Maybe, that sounds like something users might do, but then again, it would not be something for `store_log()` to decide, but for other parts of tmt, e.g `tmt.base.Run` which implements `tmt run`: this class would decide where to put logs, and it would task `provision` step with the right directory.

happz commented on tmt/steps/provision/local.py:
```diff
@@ -133,6 +133,43 @@ def pull(
+        """save log content and return the path"""
+        if log_path and log_name:
+            # if log_path contains log file name
+            if log_path.name == log_name:
```
`if log_path.is_dir` would be better - what if the caller gives us a filename that does not match the log name? It's not up to us and here to dispute that.

skycastlelily replied:
updated^^

skycastlelily commented:
Updated^^,added an Optional log_path param to handle_guest_logs,in case we
want to let users specify the log dir in the future,though users can
specify --workdir-root.

On Tue, Jan 30, 2024 at 12:08 AM Miloš Prchlík ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tmt/steps/provision/local.py
> <https://github.com/teemtee/tmt/pull/2647#discussion_r1469827525>:
>
> > @@ -133,6 +133,43 @@ def pull(
>              extend_options: Optional[list[str]] = None) -> None:
>          """ Nothing to be done to pull workdir """
>
> +    def acquire_log(self, log_name: str) -> Optional[str]:
> +        """fetch and return content of a requested log"""
> +        if log_name == 'dmesg':
> +            return self.execute(Command('dmesg')).stdout
> +        return None
> +
> +    def store_log(
> +            self,
> +            log_name: Optional[str],
> +            log_path: Optional[Path],
>
> Got it, I was thinking maybe users want to store the logs to dir they ran
> tmt command:)
>
> Maybe, that sounds like something users might do, but then again, it would
> not be something for store_log() to decide, but for other parts of tmt,
> e.g tmt.base.Run which implements tmt run: this class would decide where
> to put logs, and it would task provision step with the right directory.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/teemtee/tmt/pull/2647#discussion_r1469827525>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKFR23AR6H3AJ5E723H3M2DYQ7CNXAVCNFSM6AAAAABCL3RX52VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNBZGA2TOMRUGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -625,6 +625,11 @@ def is_ready(self) -> bool:
+    @property
+    def logs(self) -> list[str]:
+
+        raise NotImplementedError
```
Seems like returning an empty list would be a nice default implementation. There are also plugins outside of tmt repository, they might get updated much later, so it would be safer to return something rather than raising an exception.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1108,42 @@ def requires(cls) -> list['tmt.base.Dependency']:
+                if log_path:
+                    self.store_log(log_path, log_content, log_name)
+                else:
+                    self.store_log(Path.cwd(), log_content, log_name)
```
`log_path` is no longer `None`, thanks to line 1136.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1108,42 @@ def requires(cls) -> list['tmt.base.Dependency']:
+                          logs: Optional[list[str]] = None) -> None:
+        """get log content and save it to the workdir/provision/,
+        list the log dir in guests.yaml"""
+        logs = logs or []
```
It might be better to initialize with `self.logs`, right? Once it returns an empty list...

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -625,6 +625,11 @@ def is_ready(self) -> bool:
         raise NotImplementedError
 
+    @property
+    def logs(self) -> list[str]:
```
I'd suggest `log_names` rather than `logs` - `logs` suggests it's really logs, name, content, maybe even more. Since the goal is to return just a list of log names, let's name the property properly.

And please, add a docstring. Check the sources, e.g. https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/__init__.py#L1317, https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/__init__.py#L630 or https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L1207

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1108,42 @@ def requires(cls) -> list['tmt.base.Dependency']:
         return []
 
+    def acquire_log(self, log_name: str) -> Optional[str]:
+        """fetch and return content of a requested log"""
```
Docstring deserves an improvement, see e.g. https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L1988, https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L2048 or https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L2243.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1108,42 @@ def requires(cls) -> list['tmt.base.Dependency']:
+            log_path: Path,
+            log_content: str,
+            log_name: Optional[str] = None) -> None:
+        """save log content and return the path"""
```
Same as above, please, update the docstring.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -625,6 +625,12 @@ def is_ready(self) -> bool:
 
+    @property
+    def log_names(self) -> list[str]:
+        """Return name list of logs the guest could provide."""
```
Check out other one-line docstrings - if they fit on one line, there should be a space before and after the text, and no full stop at the end of sentence.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1109,59 @@ def requires(cls) -> list['tmt.base.Dependency']:
+    def acquire_log(self, log_name: str) -> Optional[str]:
+        """
+        Fetch and return content of a log.
+        :param log_name: name of the log.
```
One empty line, please. See how other docstrings look like, let's stay consistent in style.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1109,59 @@ def requires(cls) -> list['tmt.base.Dependency']:
+        """
+        Fetch and return content of a log.
+        :param log_name: name of the log.
+        :returns: content of the log.
```
The `Optional` deserves to be explained as well: when would `None` be returned? E.g. "content of the log, or `None` if the log cannot be retrieved"?

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1109,59 @@ def requires(cls) -> list['tmt.base.Dependency']:
+        """
+        Save log content to a file.
+        :param log_path: a path to save into,could be a directory
+        or a file path.
```
This line needs to be indented by 4 spaces, to be rendered correctly.

happz replied:
Also, there should be a space after `into,`.

skycastlelily replied:
By "this line",you mean "or a file path"? I failed to see the problem,I have ran "pre-commit run --all-files" before I push the code.

happz replied:
Yes, I meant "or a file path". It's what Sphinx expects, if one needs to split a lengthy parameter description, lines need to be indented to be rendered properly.

```
:param foo: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed
    do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
    enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi
    ut aliquip ex ea commodo consequat.
```

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1109,59 @@ def requires(cls) -> list['tmt.base.Dependency']:
+        or a file path.
+        :param log_content: content of the log.
+        :param log_name: name of the log, if not set, log_path
+        is supposed to be '/dev/null' or  a file path.
```
I think this special handling of `/dev/null` is a needless complication. It's a file like any other. E.g. the test below, `if not log_path.is_dir() or str(log_path) == '/dev/null':` - the check for `/dev/null` will never be performed, because `/dev/null` is not a directory, and that's enough for the condition to be true.

Let's ask a question, will there be a special branch for other magic files, like `/dev/sdb` or `/dev/zero`? :) I see no advantage, why do you think it's needed for `/dev/null`? If the user decides to not save guest logs, there would probably be an option that would tell tmt to not even call `handle_guest_logs()`. It would be much easier to not save logs at all rather than fetching them and dumping them into `/dev/null`. And as I said, it's a file like any other anyway...

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1103,6 +1109,59 @@ def requires(cls) -> list['tmt.base.Dependency']:
+            if log_name:
+                name_str = log_name
+            else:
+                name_str = 'tmt-guestlog-' + \
```
Nah, I wouldn't bother. File name or log name are inputs, it's up to the caller to provide them. If `log_path.is_dir()` is true and `log_name` is not set, let's raise an exception instead of inventing names. If there's no `log_name`, we would end up with a bunch of files with very similar names - when it was clearly a bug in the upper level that did know log names but did not bother to give them to this method.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1112,8 +1112,9 @@ def requires(cls) -> list['tmt.base.Dependency']:
+
         :param log_name: name of the log.
-        :returns: content of the log.
+        :returns: content of the log, or None if the log cannot be retrieved.
```
"\`\`None\`\`" instead of "None", to make it clear it's not just any "none", but literally Python's `None` object.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1124,34 +1125,33 @@ def store_log(
-                    datetime.datetime.now(datetime.timezone.utc).strftime("%Y%m%d%H%M%S")
-                self.warn(f'log_name is not set, using file name:{name_str}')
-            (log_path / name_str).write_text(log_content)
+                raise tmt.utils.GeneralError('log_name is None.')
```
Let's help the user a bit :) How about "Log path is a directory but log name is not defined."? Or something similar, to provide more context.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1124,34 +1125,33 @@ def store_log(
+            or Path.cwd() will be used.
         :param log_names: name list of logs need to be handled.If not set,
-        self.log_names will be used.
+            self.log_names will be used.
```
Few more notes:

* there should be a space between the end of a sentence and starting a new one, i.e. "a directory to save into. If not set...". The same applies to comas, i.e. "If not set, ..."
* "current working directory" is better than "Path.cwd()".
* "step's working directory" might be better than "self.workdir". It explains better the flow of decisions. You may use both, of course, but then you should take advantage of  Sphinx' domains: "If not set, step's working directory (:py:attr:\`workdir\`) or current working directory will be used."
* the same goes for "self.log_names" - "If not set, all guest logs will be collected." is better than "self.log_names" as it conveys the semantics of the parameter. A link to "self.log_names" might be nice, but it shouldn't be the first choice: "if not set, all guest logs would be collected, as reported by :py:attr:\`log_names\`."

happz commented on tmt/steps/provision/local.py:
```diff
@@ -141,8 +141,9 @@ def pull(
+
         :param log_name: name of the log.
-        :returns: content of the log.
+        :returns: content of the log, or None if the log cannot be retrieved.
```
Same as above, "\`None\`" to make it a literal.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1477,6 +1483,59 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
 
         return []
 
+    def acquire_log(self, log_name: str) -> Optional[str]:
```
Nomenclature comment: in tmt code the form of `[a-z]+name` is more common than `[a-z]+_name` - `filename`, `guestname, also `filepath`, etc. - therefore I would recommend using `logname` instead of `log_name`.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1477,6 +1483,59 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
+            self,
+            log_path: Path,
+            log_content: str,
+            log_name: Optional[str] = None) -> None:
```
Nomenclature comment number two: as mentioned above, `*_name` is much less common in tmt codebase, therefore I would recommend some changes to names here: `log_path` => `filepath` (we use `filepath` a lot) or just `path` since it can be also a directory, `log_content` => `content` (it will be clear what content is meant by that), and `log_name` => `logname`.

skycastlelily replied:
Addressed: https://github.com/teemtee/tmt/pull/2647/commits/a19fc72cd90340def62534c69b088fe8969b7786 ^^

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1477,6 +1483,59 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
+
+    def handle_guest_logs(self,
+                          log_path: Optional[Path] = None,
+                          log_names: Optional[list[str]] = None) -> None:
```
To be aligned with the rest, `log_path` => `path`, or maybe `dirpath` as it cannot be a file, and `log_names` should probably be `lognames`.

skycastlelily replied:
Addressed: https://github.com/teemtee/tmt/commit/a19fc72cd90340def62534c69b088fe8969b7786 ^^

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1477,6 +1483,59 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
+        log_path = log_path or self.workdir or Path.cwd()
+        for log_name in log_names:
+            log_content = self.acquire_log(log_name)
+            if log_content:
```
If content is an empty string, we should probably save it. Empty string should not be treated as "undefined" or "unknown", that's what we have `None` for, and `acquire_log()` will return `None` if the log is meaningless or something. An empty string may be completely valid content of a log, and we should save it. Whether it's a good sign or a bad sign, that would be up to the user who would see an expected file, e.g. `messages`, but it would be empty.

skycastlelily replied:
Addressed: https://github.com/teemtee/tmt/commit/a19fc72cd90340def62534c69b088fe8969b7786 ^^

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1477,6 +1483,59 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
+            raise tmt.utils.GeneralError(
+                'Log path is a directory but log name is not defined.')
+
+    def handle_guest_logs(self,
```
I'd try to find a better name, "handle guest logs" may mean a lot of different things. How about `fetch_guest_logs()`? BTW it's a method of `Guest` class, I think we can drop the `guest` bit, `Guest.fetch_logs()` will be pretty clear.

skycastlelily replied:
Addressed: https://github.com/teemtee/tmt/commit/a19fc72cd90340def62534c69b088fe8969b7786 ^^

happz commented:
As briefly discussed on Slack, there might be a better implementation, bundling together log name & a way to fetch it in a class. That should give us better handling and responsibilities, plus making nicer spot for logs that do not need to be fetched at all, like https://github.com/teemtee/tmt/pull/3511

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -993,6 +1008,8 @@ def get_data_class(cls) -> type[GuestData]:
     #: Guest topology hostname or IP address for guest/guest communication.
     topology_address: Optional[str] = None
 
+    guest_logs: list[GuestLog] = []
```
1. it's already in `Guest` class, `guest_` prefix is superfluous. `Guest.guest_logs` vs `Guest.logs`
2. a a class-level mutable list, I would be worried that `self.guest_logs.append()` below would add guest-specific logs to one shared list. The list should probably be initialized in `__init__()` instead.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1646,6 +1663,60 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
 
         return []
 
+    def acquire_log(self, logname: str) -> Optional[str]:
```
This method is probably not needed anymore.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1646,6 +1663,60 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
+        """
+        raise NotImplementedError
+
+    def store_log(self, path: Path, content: str, logname: Optional[str] = None) -> None:
```
This method can move to `GuestLog` class, right? There's `log.fetch()` followed by `store_log()`, why not have e.g. `save()` method accepting this arguments, which would then call `self.fetch()` and `self.store()`?

happz commented on tmt/steps/provision/mrack.py:
```diff
@@ -1458,3 +1463,22 @@ def guest(self) -> Optional[GuestBeaker]:
+
+        if self.name == 'dmesg':
+            return self.guest.execute(Command('dmesg')).stdout
+        return tmt.utils.get_url_content(self.url) if self.url else None
```
1 `dmesg` should have its own subclass, right? It has a different implementation of `fetch()` after all
2. I'm not sure it would be always possible to run `self.guest.execute()`, the guest may be dead. It would be better to download Beaker's `console.log` instead.

happz commented on tmt/steps/provision/mrack.py:
```diff
@@ -1458,3 +1463,22 @@ def guest(self) -> Optional[GuestBeaker]:
+
+@container
+class GuestLogBeaker(tmt.steps.provision.GuestLog):
+    def __init__(self, guest: GuestBeaker, name: str, url: Optional[str] = None) -> None:
```
With `container`, you shouldn't need custom `__init__()`, one will be generated automagically for you.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1646,6 +1663,60 @@ def essential_requires(cls) -> list['tmt.base.Dependency']:
+
+        guest_logs = guest_logs or self.guest_logs or []
+
+        dirpath = dirpath or (self.workdir / 'logs' if self.workdir else None) or Path.cwd()
```
There should be a `logdir` property of `Guest` class that does this: every part of code interested in finding where guest's logs are should use `guest.logdir` instead of constructing the part on their own. Then you should be able to simply run `guest.logdir.mkdir(parents=True, exists_ok=True)` instead of checking `self.workdir` again below.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -2642,6 +2713,28 @@ def show(self, keys: Optional[list[str]] = None) -> None:
+        if self.workdir is None:
+            return
+
+        logs_dir = self.workdir / 'logs'
```
Another point: trying to provide custom implementation to preserve some (dynamic) filenames, and by doing so, we get different levels of logging and different behavior from one step only. That's not good. Maybe we should first change how `Step.prune()` works, and instead of file/directory names, we should use regular expressions. Then we could just add `logs/.*\.txt` to `Provision._preserved_workdir_members` and keep using the default `prune()` method. That would be a nice, self-contained PR right there ;)

skycastlelily replied:
I guess the problem is  plugin.prune will remove the whole workdir even if there are preserved members in it:
and here is the mr to solve it:https://github.com/teemtee/tmt/pull/3548^^

skycastlelily commented:
wow,thanks for your super quick response and valuable feedbacks,impressed again,and updated^^

happz commented:
This PR should **not** have been closed :(

@skycastlelily can you please restart the PR, probably by creating a new one?

skycastlelily commented:
Again?! :rofl: 
I should keep in mind: never  rebase on  a likely being merged-soon branch next time :upside_down_face:  
Here is the new mr:https://github.com/teemtee/tmt/pull/3602

happz commented:
> Again?! 🤣 I should keep in mind: never  rebase on  a likely being merged-soon branch next time 🙃 Here is the new mr:#3602

Yes, again. it was probably mentioned in the merged PR, in a sentence like "this would be resolved by that PR", and GitHub closed it for us...
